feat(docs): cheatcode docs

rahul-kothari opened the pull request:
fIX #1458

rahul-kothari committed:
add cheatcode docs

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+// any noir contract calls that make use of current timestamp now will have `newTimestamp`
+```
+
+### loadPublic
````
We should probably add a comment here related to structs as they are not in a single slot, e.g., if your example were using a struct the `loadPublic` would only load part of it.

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,449 @@
+const newTimestamp = timestamp + 100_000_000;
+await cc.l2.warp(newTimestamp);
+
+// any noir contract calls that make use of current timestamp now will have `newTimestamp`
```
```suggestion
// any noir contract calls that make use of current timestamp and is executed in the next rollup block will now read `newTimestamp`
```

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+```
+
+#### Description
+Set time of the next execution on L2 and L1 for next execution.
````
```suggestion
Sets the time of L1 and the time of the next L2 block.
```

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,449 @@
+
+#### Description
+Set time of the next execution on L2 and L1 for next execution.
+Like with the corresponding L1 cheatcode, time can only be set in the future, not the past.
```
Will the cheatcode throw an error if called in the past, or will things simply break? If they are just breaking, better to throw explicitly.

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,449 @@
+import { createAztecRpcClient, CheatCodes } from '@aztec/aztec.js';
+const aztecRpcUrl = 'http://localhost:8080';
+const aztecRpcClient = createAztecRpcClient(aztecRpcUrl);
+const cc = CheatCodes.create(aztecRpcUrl, aztecRpcClient);
```
```suggestion
const cc = await CheatCodes.create(aztecRpcUrl, aztecRpcClient);
```

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,449 @@
+- Aztec network.
+### Why is this useful?
+
+<!-- Contextualise why a dapp developer needs this. What use cases / products / features does this unlock? Any real world examples? -->
```
You should probably create an issue for this such that it is not forgotten or finish it as part of this pr.

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,449 @@
+There are two properties of the CheatCodes class - `l1` and `l2` for cheatcodes relating to the L1 and L2 (Aztec) respectively.
+
+## L1 related cheatcodes
+These are cheatcodes exposed from anvil. But instead of having to know the right RPC calls to make and format the curl request appropirately, there are convenient wrappers exposed to developers.
```
```suggestion
These are cheatcodes exposed from anvil conveniently wrapped for ease of use in the Sandbox.
```

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,449 @@
+// Set the timestamp for the next block on L1. 
+public async setNextBlockTimestamp(timestamp: number): Promise<void> 
+
+// Dumps the current L1 chain state to a given file. Can be used with `loadChainState()` to first create a forked version of mainnet on anvil (to for example interact with uniswap), save the chain state to a file, and later load it to the sandbox' anvil instance.
```
Should probably be multiple lines, if seen in most browsers will move out of the field of view.

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,449 @@
+
+#### Description
+Sets the timestamp (unix format in seconds) for the next mined block on L1  
+Remember that timestamp can only be set in the future and not in the past.
```
Will the cheatcode throw an error if called in the past, or will things simply break? If they are just breaking, better to throw explicitly.

rahul-kothari replied:
throws an explicit error

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+
+#### Example
+```js
+await cc.l1.setNextBlockTimestamp(1692183270) // Set next block timestamp to 16 Aug 2023 10:54:30 GMT
````
The comment here will most often go out of the field of view. Probably better to have on the line above etc.

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+```
+
+#### Description
+Set the bytecode for an L1 contract
````
```suggestion
Set the bytecode for an L1 contract.
```

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+```
+
+#### Description
+Mines the specified number of blocks on L1 (default 1)
````
```suggestion
Mines the specified number of blocks on L1 (default 1).
```

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+```
+
+#### Description  
+Loads the value at a storage slot of a L1 contract
````
```suggestion
Loads the value at a storage slot of a L1 contract.
```

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+```
+
+#### Description
+Stores the value in storage slot on a L1 contract
````
```suggestion
Stores the value in storage slot on a L1 contract.
```

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+```
+
+#### Description 
+Computes the storage slot for a map key
````
```suggestion
Computes the storage slot for a map key.
```

LHerskind reviewed (commented):
Nits and grits

rahul-kothari committed:
Apply suggestions from code review

Co-authored-by: Lasse Herskind <16536249+LHerskind@users.noreply.github.com>

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,449 @@
+
+Most of the time, simply testing your smart contract isn't enough. To manipulate the state of the Aztec blockchain, as well as test for specific reverts and events, the sandbox is shipped with a set of cheatcodes.
+
+Cheatcodes allow you to change the time of the L2 block, load certain state or more easily manipulate L1 instead of having to write dedicated RPC calls to anvil. 
```
This has just made me think, should we make the cheat codes agnostic of hardhat / anvil

rahul-kothari replied:
hmm aztec uses anvil within the sandbox though

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+### Initialisation
+```js
+import { createAztecRpcClient, CheatCodes } from '@aztec/aztec.js';
+const aztecRpcUrl = 'http://localhost:8080';
````
The create signature is 
create(ethRpcUrl, aztecRpcUrl)

rahul-kothari replied:
Doesn't look like it?
For creating the aztec rpc client, the sandbox uses `createAztecRpcClient(aztecRpcUrl);` [here](https://github.com/AztecProtocol/aztec-packages/blob/4af4845f80b1be548efa1ca79f5588bb1c7f1423/yarn-project/aztec-sandbox/src/examples/uniswap_trade_on_l1_from_l2.ts#L58)

And for cheatcodes it is `create(rpcUrl: string, aztecRpc: AztecRPC)`

Maddiaa0 replied:
My mistake, youre absolutely correct

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,449 @@
+const cc = await CheatCodes.create(aztecRpcUrl, aztecRpcClient);
+```
+
+There are two properties of the CheatCodes class - `l1` and `l2` for cheatcodes relating to the L1 and L2 (Aztec) respectively.
````
This has made me think that the interface should change to `.eth` and `.aztec` rather than `l1` and `l1`. 
This would make it a good bit more explicit

rahul-kothari replied:
I actually really like this - will modify!

LHerskind replied:
I think is better. Have approved the change in the #1611.

Maddiaa0 reviewed (commented):
temp review before pushing changes

rahul-kothari committed:
prank->impersonate, add throw tests, respond to PR comments

rahul-kothari committed:
use .eth and .aztec instead of .l1 and .l2 (#1611)

In case we like the `.eth` and `.aztec` variation over `.l1` and `.l2`
in cheatcodes!

This would get merged into my docs pr #1585

rahul-kothari committed:
use hardhat_*

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,458 @@
+
+#### Description
+Sets the timestamp (unix format in seconds) for the next mined block on Ethereum.
+Remember that timestamp can only be set in the future and not in the past.
```
nit  its probably worth being explicit about the erroring

LHerskind replied:
Probably related to: https://github.com/AztecProtocol/aztec-packages/pull/1585#discussion_r1297072567

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,458 @@
+Stores the value in storage slot on a Ethereum contract.
+
+#### Example
+```
````
probably worth annotating the solidity contracts with `sol and using separate code blocks

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,458 @@
+- @aztec/aztec.js
+
+### Initialisation
+```js
````
should get better highlighting using `ts rather than js

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,458 @@
+These are cheatcodes exposed from anvil/hardhat conveniently wrapped for ease of use in the Sandbox.
+
+### Interface
+```
````
annotate `ts

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,458 @@
+These are cheatcodes specific to manipulating the state of Aztec rollup.
+
+### Interface
+```
````
annotate ts

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,458 @@
+
+#### Example  
+```
+/// struct Storage {
````
noir snippets can be annotated rust for better results

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,458 @@
+Stores the value in storage slot on a Ethereum contract.
+
+#### Example
+```
+/// contract LeetContract {
+///     uint256 private leet = 1337; // slot 0
+/// }
+
+const leetContractAddress = EthAddress.fromString('0x1234...');
+await cc.eth.store(leetContractAddress, BigInt(0), BigInt(1000));
+const value = await cc.eth.load(leetContractAddress, BigInt(0));
+console.log(value); // 1000
+```
````
```solidity
contract LeetContract {
     uint256 private leet = 1337; // slot 0
 }
```
```ts
const leetContractAddress = EthAddress.fromString('0x1234...');
await cc.eth.store(leetContractAddress, BigInt(0), BigInt(1000));
const value = await cc.eth.load(leetContractAddress, BigInt(0));
console.log(value); // 1000
```

Solidity blocks can be annotated with solidity, they arent perfect in docusaurus but it does offer some highlighting.

Maddiaa0 reviewed (commented):
Overall great content, however noir snippets can be annotated with `rust, the ones currently js should be ts to get better highlighting and solidity ones can be separated rather than inline commented and annotated with `solidity

rahul-kothari committed:
use better syntax highlighting

rahul-kothari committed:
explicit error

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
````diff
@@ -0,0 +1,468 @@
+#### Example  
+```rust
+struct Storage {
+    // highlight-next-line:PublicState
````
is this still required

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,468 @@
+#### Description
+Loads the value stored at the given slot in the public storage of the given contract. 
+
+Note: One Field element occupies a storage slot. So structs with multiple field elements won't fit in a single slot. So using loadPublic would only load a part of the struct (depending on the size of the attributes within it).
```
```suggestion
Note: One Field element occupies a storage slot. Hence, structs with multiple field elements will be spread over multiple slots, rather they will be spread over multiple sequential slots. Using loadPublic will only load a single field of the struct (depending on the size of the attributes within it).
```
Id reword this slightly

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,468 @@
+
+Keep up with the latest discussion and join the conversation in the [Aztec forum](https://discourse.aztec.network).
+
+You can also use the above link to request for more cheatcodes.
```
```suggestion
You can also use the above link to request more cheatcodes.
```

Maddiaa0 commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,468 @@
+#### Description
+Dumps the current Ethereum chain state to a file.
+Stores a hex string representing the complete state of the chain in a file with the provided path. Can be re-imported into a fresh/restarted instance of Anvil to reattain the same state.
+When combined with `loadChainState()` cheatcode, it can be let you easily import the current state of mainnet into the Anvil instance of the sandbox, to use Uniswap for example.
```
The uniswap example here is really specific to our testing case and probably confusing without context. Lets drop it, people looking for this feature likely have in mind a specific use case

rahul-kothari committed:
address sean's nits

rahul-kothari committed:
improve struct slot wording

LHerskind commented on docs/docs/dev_docs/testing/cheat_codes.md:
```diff
@@ -0,0 +1,466 @@
+#### Description
+Dumps the current Ethereum chain state to a file.
+Stores a hex string representing the complete state of the chain in a file with the provided path. Can be re-imported into a fresh/restarted instance of Anvil to reattain the same state.
+When combined with `loadChainState()` cheatcode, it can be let you easily import the current state of mainnet into the Anvil instance of the sandbox.
```
The wording is a little weird here, `it can be let you` and it might not be the `current` state. Maybe fine to point to the other cheatcode and say use in combination.

```suggestion
To be used in combination with `loadChainState()`
```
